### PR TITLE
#11278: Fix compilation issues on DeviceMeshView for GCC-13

### PR DESCRIPTION
**Pull request by cfjchu**
This fixes compilation issues on GCC-13 reported and patched by @marty1885.

### Ticket
[Link to Github Issue
](https://github.com/tenstorrent/tt-metal/issues/11278)

### Problem description
Compilation issues on GCC-13 reported and patched by @marty1885 on DeviceMeshView class.

### What's changed
- Partial patch provided by @marty1885 submitted to resolve compilation issue on GCC-13. 

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by marty1885** [comment]
@cfjchu 

I am pretty sure this is not the case. Still commenting to be sure. My patches is not sufficient to get tt-metal compiling on GCC. It only gets the compilation far enough to run into issues that only TT can solve. More patches will be needed to have tt-metal working on GCC.

File: tt_metal/impl/device/device_mesh_view.cpp [hunk 1]
```diff
@@ -76,7 +76,7 @@ DeviceMeshView::DeviceView DeviceMeshView::get_devices(const Coordinate& start,
 }
 
 DeviceMeshView::DeviceView DeviceMeshView::get_devices(const DeviceGrid& shape) {
-    return get_devices({0, 0}, {shape.first - 1, shape.second - 1});
+    return get_devices({0, 0}, {(size_t)(shape.first - 1), (size_t)(shape.second - 1)});
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/device_mesh_view.cpp:79]: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast

**Review by ayerofieiev-tt — APPROVED** [review]

**Comment by cfjchu** [comment]
I spoke with @ayerofieiev-tt and @TT-billteng to get better context. I'm going to close this PR without merging this. While it fixes _some_ of the compile errors, the intention for this PR wasn't to support full GCC-13 compilation. It seems one of the workarounds to https://github.com/tenstorrent/tt-metal/issues/10652 was to compile with C++23 but we should not be relying on any C++23 features in our code. We're working on getting gcc on CI and the plan is to fix any compilation issues as part of those changes. This is being tracked here: https://github.com/tenstorrent/tt-metal/issues/10652

**Comment by marty1885** [comment]
@cfjchu @ayerofieiev-tt Is it possible to have a short term solution that makes tt-metal compile on GCC? My GGML backend for tt-metal is blocked due to GGML depending on some system C++ packages. If not I will attempt to hack tt-mertal so it works against libstdc++.

**Comment by TT-billteng** [comment]
> @cfjchu @ayerofieiev-tt Is it possible to have a short term solution that makes tt-metal compile on GCC? My GGML backend for tt-metal is blocked due to GGML depending on some system C++ packages. If not I will attempt to hack tt-mertal so it works against libstdc++.

I've just checked in a 22.04 dockerfile, @cfjchu is fixing the compile errors

The [branch](https://github.com/tenstorrent/tt-metal/commits/bteng/build_2204) adds g++-12 compilation into CI testing and can be merged whenever @cfjchu is done

**Comment by marty1885** [comment]
Will try that. Thank you!

**Comment by marty1885** [comment]
FYI. So far the `bteng/build_2204` branch runs into the same `static_assert(sizeof(BaseType) <= MAX_STORAGE_SIZE);` error on GCC-13. Also I think a new error:

```
/home/marty/Documents/tt-metal/ttnn/cpp/ttnn/decorators.hpp:309:77: error: friend declaration ‘consteval auto ttnn::decorators::get(operation_name_key_t<cpp_fully_qualified_name>)’ declares a non-template function [-Werror=non-template-friend]
  309 |     friend consteval auto get(operation_name_key_t<cpp_fully_qualified_name>);
```

**Pull request closed without merging**